### DISCUSSION TITLE
Updates to the linearity algorithm spec

### DISCUSSION THREAD

**Discussion by crabmusket** [austral/austral#522] [General]
I really love Austral's goal of implementing a fits-in-head sized algorithm for linearity checking (rather than piling on heuristics and compiler machinery). I also love the commitment to having a spec (rather than "the language is what the compiler does"). The current linearity checking spec [reads](https://austral-lang.org/spec/spec.html#linearity):

> 1. A type is linear if it physically contains a linear type, or is declared to be linear (belongs to either the `Linear` or `Type` universes).
> 2. A variable `x` of a linear type must be used once and exactly once in the scope in which it is defined, where "used once" means:
>    1. If the scope is a single block with no changes in control flow, a variable that appears once anywhere in the block is used once.
>    2. If the scope includes an `if` statement, the variable must be used once in every branch or it must not appear in the statement.
>    3. Analogously, if the scope includes a `case` statement, the variable must be used once in every `when` clause, or it must not appear in the statement.
>    4. If the scope includes a loop, the variable may not appear in the loop body or the loop start/end/conditions (because it would be used once for each iteration). Linear variables can be _defined_ inside loop bodies, however.
>    5. A borrow statement does not count as using the variable.
>    6. Neither does a path that ends in a free value (this is the one concession to programmer ergonomics).
> 3. When a variable is used, it is said to be consumed. Afterwards, it is said to be in the consumed state.
> 4. If a variable is never consumed, or is consumed more than once, the compiler signals an error.
> 5. Linear values which are not stored in variables cannot be discarded.
> 6. You cannot return from a function while there are any unconsumed linear values.
> 7. A borrowed variable cannot appear in the body of the statement that borrows it.

I'm aware of one way the compiler currently deviates from the text of the spec: linear variables can be used in loop bodies as long as you "put them back" (#463). As a side effect of this change, it seems the same thing can be done in if and case statements (#519). This implies the spec needs to change!

It sounds like @eudoxia0 has a generalised form of the linearity algorithm in mind already. I'm sure it will improve on what I wanted to suggest, but I hope there's some value writing this out anyway. I have a few suggestions for the spec:

- Explicitly list what "uses" a variable:
    - Returning it
    - Passing it as an argument to a function or constructor
    - Destructuring it
    - Moving it to another variable
- Reword the rules for `if` something along the lines of: "a variable must be _consumed_ in either all or none of the branches of an `if` statement". (I don't think this is rigorous enough yet, and it only goes halfway to allowing consume/replace.)
- Instead of special cases for `while` and `case` statements, I'd suggest handling them as syntax sugar over `if` and `while true`. This reduces the rules to scope, branch and syntax de-sugaring.

For an example of reducing loops to syntax sugar, the following rule could apply:

```
while $condition do
    $body
end while;
```

Is equivalent to:

```
while true do
    if not $condition then
        break;
    end if;
    $body
end while
```

Now there is no need to specify "linear variables cannot be consumed in the condition of a `while` loop", because of the rule about conditional branches.

(Austral does not have `break`, so this may need to be described as some kind of linearity-checking intrinsic? It only has a chance of being simple because programmers understand how break works from other languages.)

I would also suggest, for the sake of the text, that comments and corollaries be moved to a section outside the core algorithm spec. For example, "You cannot return from a function while there are any unconsumed linear values" is useful for the reader of the spec, since it can confirm their intuition or understanding of the algorithm. But it is redundant with the "must be used once and exactly once" stipulation.

**Comment by eudoxia0** [comment]
Yes, I have to update the spec. I think I want the description of the algorithm to be close to the source and avoid desugaring. Rather, I think it might be useful to have multiple descriptions of the algorithm.

1. A description like what's in the spec right now: that walks the code in depth-first/execution order, analogous to how people read code.
2. A "minimal" description that drops some rules by describing how code can be desugared first.
3. A graph-based description where you transform the code into a control flow graph, and describe the linearity rules on top of the graph structure.
